File: CS/RefreshHelperClass.cs
using System;
using System.Collections;
using DevExpress.XtraGrid;
using DevExpress.Utils;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraGrid.Views.Grid;

namespace DevExpress.XtraGrid.Helpers {
	public class RefreshHelper {
		[Serializable]
		public struct RowInfo{
			public object Id;
			public int level;
		};

		private GridView view;
		private string keyFieldName;
		private ArrayList saveExpList;
		private ArrayList saveSelList;
		private ArrayList saveMasterRowsList;
		private int visibleRowIndex = -1;

		public RefreshHelper(GridView view, string keyFieldName) {
			this.view = view;
			this.keyFieldName = keyFieldName;
		}

		public ArrayList SaveExpList{
			get { 
				if(saveExpList == null)
					saveExpList = new ArrayList();
				return saveExpList;
			}
		}

		public ArrayList SaveSelList{
			get { 
				if(saveSelList == null)
					saveSelList = new ArrayList();
				return saveSelList;
			}
		}

		public ArrayList SaveMasterRowsList{
			get{
				if(saveMasterRowsList == null)
					saveMasterRowsList = new ArrayList();
				return saveMasterRowsList;
			}
		}

		protected int FindParentRowHandle(RowInfo rowInfo, int rowHandle) {
			int result = view.GetParentRowHandle(rowHandle);
			while(view.GetRowLevel(result) != rowInfo.level && view.IsValidRowHandle(result))
				result = view.GetParentRowHandle(result);
			return result;
		}

		protected void ExpandRowByRowInfo(RowInfo rowInfo){
			int dataRowHandle = view.LocateByValue(0, view.Columns[keyFieldName], rowInfo.Id);
			if(dataRowHandle != GridControl.InvalidRowHandle){
				int parentRowHandle = FindParentRowHandle(rowInfo, dataRowHandle);
				view.SetRowExpanded(parentRowHandle, true, false);
			}
		}

		protected int GetRowHandleToSelect(RowInfo rowInfo) {
			int dataRowHandle = view.LocateByValue(0, view.Columns[keyFieldName], rowInfo.Id);
			if(dataRowHandle != GridControl.InvalidRowHandle)
				if (view.GetRowLevel(dataRowHandle) != rowInfo.level)
					return  FindParentRowHandle(rowInfo, dataRowHandle);
			return dataRowHandle;
		}

		protected void SelectRowByRowInfo(RowInfo rowInfo, bool isFocused){
			if(isFocused)
				view.FocusedRowHandle = GetRowHandleToSelect(rowInfo);
			else 
				view.SelectRow(GetRowHandleToSelect(rowInfo));
		}

		public void SaveSelectionViewInfo(ArrayList list){
			list.Clear();
			GridColumn column = view.Columns[keyFieldName];
			RowInfo rowInfo;
			int[] selectionArray = view.GetSelectedRows();
			if (selectionArray != null)  // otherwise we have a single focused but not selected row
				for(int i = 0; i < selectionArray.Length; i++){
					int dataRowHandle = selectionArray[i];
					rowInfo.level = view.GetRowLevel(dataRowHandle);
					if(dataRowHandle < 0) // group row
						dataRowHandle = view.GetDataRowHandleByGroupRowHandle(dataRowHandle);
					rowInfo.Id = view.GetRowCellValue(dataRowHandle, column);            
					list.Add(rowInfo);
				}
			rowInfo.Id = view.GetRowCellValue(view.FocusedRowHandle, column);
			rowInfo.level = view.GetRowLevel(view.FocusedRowHandle);
			list.Add(rowInfo);
		}

		public void SaveExpansionViewInfo(ArrayList list){
			if(view.GroupedColumns.Count == 0) return;
			list.Clear();
			GridColumn column = view.Columns[keyFieldName];
			for(int i = -1; i > int.MinValue; i--) {
				if(!view.IsValidRowHandle(i)) break;
				if(view.GetRowExpanded(i)){
					RowInfo rowInfo;
					int dataRowHandle = view.GetDataRowHandleByGroupRowHandle(i);
					rowInfo.Id = view.GetRowCellValue(dataRowHandle, column);
					rowInfo.level = view.GetRowLevel(i);
					list.Add(rowInfo);
				}			  
			}			 
		}

		public void SaveExpandedMasterRows(ArrayList list){
			if(view.GridControl.Views.Count == 1) return;
			list.Clear();
			GridColumn column = view.Columns[keyFieldName];
			for(int i = 0; i < view.DataRowCount; i++)
				if(view.GetMasterRowExpanded(i))
					list.Add(view.GetRowCellValue(i, column));
		}

		public void SaveVisibleIndex(){
			visibleRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - view.TopRowIndex;
		}

		public void LoadVisibleIndex(){
			view.MakeRowVisible(view.FocusedRowHandle, true);
			view.TopRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - visibleRowIndex;
		}

		public void LoadSelectionViewInfo(ArrayList list) {
			view.BeginSelection();
			try {
				view.ClearSelection();
				for(int i = 0; i < list.Count; i++)
					SelectRowByRowInfo((RowInfo)list[i], i == list.Count - 1);
			}
			finally {
				view.EndSelection();
			}
		}

		public void LoadExpansionViewInfo(ArrayList list) {
			if(view.GroupedColumns.Count == 0) return;
			view.BeginUpdate();
			try {
				view.CollapseAllGroups();
				foreach(RowInfo info in list)
					ExpandRowByRowInfo(info);
			}
			finally {
				view.EndUpdate();
			}
		}

		public void LoadExpandedMasterRows(ArrayList list) {
			view.BeginUpdate();
			try {
				view.CollapseAllDetails();
				GridColumn column = view.Columns[keyFieldName];
				for(int i = 0; i < list.Count; i++){
					int rowHandle = view.LocateByValue(0, column, list[i]);
					view.SetMasterRowExpanded(rowHandle, true);
				}
			}
			finally {
				view.EndUpdate();
			}
		}

		public void SaveViewInfo(){
			SaveExpandedMasterRows(SaveMasterRowsList);
			SaveExpansionViewInfo(SaveExpList);
			SaveSelectionViewInfo(SaveSelList);
			SaveVisibleIndex();
		}

		public void LoadViewInfo(){
			LoadExpandedMasterRows(SaveMasterRowsList);
			LoadExpansionViewInfo(SaveExpList);
			LoadSelectionViewInfo(SaveSelList);
			LoadVisibleIndex();
		}
	}
}

File: VB/RefreshHelperClass.vb
Imports System
Imports System.Collections
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraGrid.Views.Grid

Namespace DevExpress.XtraGrid.Helpers

    Public Class RefreshHelper

        <Serializable>
        Public Structure RowInfo

            Public Id As Object

            Public level As Integer
        End Structure

        Private view As GridView

        Private keyFieldName As String

        Private saveExpListField As ArrayList

        Private saveSelListField As ArrayList

        Private saveMasterRowsListField As ArrayList

        Private visibleRowIndex As Integer = -1

        Public Sub New(ByVal view As GridView, ByVal keyFieldName As String)
            Me.view = view
            Me.keyFieldName = keyFieldName
        End Sub

        Public ReadOnly Property SaveExpList As ArrayList
            Get
                If saveExpListField Is Nothing Then saveExpListField = New ArrayList()
                Return saveExpListField
            End Get
        End Property

        Public ReadOnly Property SaveSelList As ArrayList
            Get
                If saveSelListField Is Nothing Then saveSelListField = New ArrayList()
                Return saveSelListField
            End Get
        End Property

        Public ReadOnly Property SaveMasterRowsList As ArrayList
            Get
                If saveMasterRowsListField Is Nothing Then saveMasterRowsListField = New ArrayList()
                Return saveMasterRowsListField
            End Get
        End Property

        Protected Function FindParentRowHandle(ByVal rowInfo As RowInfo, ByVal rowHandle As Integer) As Integer
            Dim result As Integer = view.GetParentRowHandle(rowHandle)
            While view.GetRowLevel(result) <> rowInfo.level AndAlso view.IsValidRowHandle(result)
                result = view.GetParentRowHandle(result)
            End While

            Return result
        End Function

        Protected Sub ExpandRowByRowInfo(ByVal rowInfo As RowInfo)
            Dim dataRowHandle As Integer = view.LocateByValue(0, view.Columns(keyFieldName), rowInfo.Id)
            If dataRowHandle <> GridControl.InvalidRowHandle Then
                Dim parentRowHandle As Integer = FindParentRowHandle(rowInfo, dataRowHandle)
                view.SetRowExpanded(parentRowHandle, True, False)
            End If
        End Sub

        Protected Function GetRowHandleToSelect(ByVal rowInfo As RowInfo) As Integer
            Dim dataRowHandle As Integer = view.LocateByValue(0, view.Columns(keyFieldName), rowInfo.Id)
            If dataRowHandle <> GridControl.InvalidRowHandle Then
                If view.GetRowLevel(dataRowHandle) <> rowInfo.level Then Return FindParentRowHandle(rowInfo, dataRowHandle)
            End If

            Return dataRowHandle
        End Function

        Protected Sub SelectRowByRowInfo(ByVal rowInfo As RowInfo, ByVal isFocused As Boolean)
            If isFocused Then
                view.FocusedRowHandle = GetRowHandleToSelect(rowInfo)
            Else
                view.SelectRow(GetRowHandleToSelect(rowInfo))
            End If
        End Sub

        Public Sub SaveSelectionViewInfo(ByVal list As ArrayList)
            list.Clear()
            Dim column As GridColumn = view.Columns(keyFieldName)
            Dim rowInfo As RowInfo
            Dim selectionArray As Integer() = view.GetSelectedRows()
            If selectionArray IsNot Nothing Then ' otherwise we have a single focused but not selected row
                For i As Integer = 0 To selectionArray.Length - 1
                    Dim dataRowHandle As Integer = selectionArray(i)
                    rowInfo.level = view.GetRowLevel(dataRowHandle)
                    If dataRowHandle < 0 Then dataRowHandle = view.GetDataRowHandleByGroupRowHandle(dataRowHandle) ' group row
                    rowInfo.Id = view.GetRowCellValue(dataRowHandle, column)
                    list.Add(rowInfo)
                Next
            End If

            rowInfo.Id = view.GetRowCellValue(view.FocusedRowHandle, column)
            rowInfo.level = view.GetRowLevel(view.FocusedRowHandle)
            list.Add(rowInfo)
        End Sub

        Public Sub SaveExpansionViewInfo(ByVal list As ArrayList)
            If view.GroupedColumns.Count = 0 Then Return
            list.Clear()
            Dim column As GridColumn = view.Columns(keyFieldName)
            For i As Integer = -1 To Integer.MinValue + 1 Step -1
                If Not view.IsValidRowHandle(i) Then Exit For
                If view.GetRowExpanded(i) Then
                    Dim rowInfo As RowInfo
                    Dim dataRowHandle As Integer = view.GetDataRowHandleByGroupRowHandle(i)
                    rowInfo.Id = view.GetRowCellValue(dataRowHandle, column)
                    rowInfo.level = view.GetRowLevel(i)
                    list.Add(rowInfo)
                End If
            Next
        End Sub

        Public Sub SaveExpandedMasterRows(ByVal list As ArrayList)
            If view.GridControl.Views.Count = 1 Then Return
            list.Clear()
            Dim column As GridColumn = view.Columns(keyFieldName)
            For i As Integer = 0 To view.DataRowCount - 1
                If view.GetMasterRowExpanded(i) Then list.Add(view.GetRowCellValue(i, column))
            Next
        End Sub

        Public Sub SaveVisibleIndex()
            visibleRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - view.TopRowIndex
        End Sub

        Public Sub LoadVisibleIndex()
            view.MakeRowVisible(view.FocusedRowHandle, True)
            view.TopRowIndex = view.GetVisibleIndex(view.FocusedRowHandle) - visibleRowIndex
        End Sub

        Public Sub LoadSelectionViewInfo(ByVal list As ArrayList)
            view.BeginSelection()
            Try
                view.ClearSelection()
                For i As Integer = 0 To list.Count - 1
                    SelectRowByRowInfo(CType(list(i), RowInfo), i = list.Count - 1)
                Next
            Finally
                view.EndSelection()
            End Try
        End Sub

        Public Sub LoadExpansionViewInfo(ByVal list As ArrayList)
            If view.GroupedColumns.Count = 0 Then Return
            view.BeginUpdate()
            Try
                view.CollapseAllGroups()
                For Each info As RowInfo In list
                    ExpandRowByRowInfo(info)
                Next
            Finally
                view.EndUpdate()
            End Try
        End Sub

        Public Sub LoadExpandedMasterRows(ByVal list As ArrayList)
            view.BeginUpdate()
            Try
                view.CollapseAllDetails()
                Dim column As GridColumn = view.Columns(keyFieldName)
                For i As Integer = 0 To list.Count - 1
                    Dim rowHandle As Integer = view.LocateByValue(0, column, list(i))
                    view.SetMasterRowExpanded(rowHandle, True)
                Next
            Finally
                view.EndUpdate()
            End Try
        End Sub

        Public Sub SaveViewInfo()
            SaveExpandedMasterRows(SaveMasterRowsList)
            SaveExpansionViewInfo(SaveExpList)
            SaveSelectionViewInfo(SaveSelList)
            SaveVisibleIndex()
        End Sub

        Public Sub LoadViewInfo()
            LoadExpandedMasterRows(SaveMasterRowsList)
            LoadExpansionViewInfo(SaveExpList)
            LoadSelectionViewInfo(SaveSelList)
            LoadVisibleIndex()
        End Sub
    End Class
End Namespace
